### ISSUE TITLE
Hydra & Weights and Biases

### ISSUE THREAD

**Issue by diegomrodrigues2** [diegomrodrigues2/py_distributed_database#171]
**História de Usuário – Semana 2: Configuração Central**

> **Como** pesquisador de LLMs,
> **quero** ter um sistema de configuração unificado usando Hydra, utilitários de logging para Weights & Biases e um pipeline de pré-processamento de dados coerente,
> **para que** todos os meus scripts possam compartilhar configurações, registrar execuções de forma padronizada e alimentar corretamente o modelo Qwen3 sem duplicar código.

---

### 2.1 Configuração com Hydra

> **Objetivo:** Gerenciar **todas** as configurações de experimento (modelo, dados, treino, avaliação) de forma centralizada e reproduzível.&#x20;

1. **Criar diretório `conf/`**

   * Sub‐pastas: `conf/model/`, `conf/data/`, `conf/train/`, `conf/eval/`.
   * Arquivo principal `conf/default.yaml` importando opções de cada grupo.
2. **Definir `default.yaml`**

   * Exemplo de hierarquia:

     ```yaml
     defaults:
       - model: qwen3_base
       - data: sft
       - train: base
       - eval: default
     ```
3. **Testar execução básica e multirun**

   * Script de exemplo (`main.py`) que usa `@hydra.main(config_path="conf", config_name="default")` e imprime os parâmetros lidos.
   * **Acceptance Criteria:**

     * `python main.py` exibe o conteúdo estruturado de `default.yaml`.
     * `python main.py -m model=qwen3_large,model=qwen3_small` dispara dois runs distintos (multirun) sem erros.&#x20;

---

### 2.2 Wrapper de Integração W\&B (`utils/wandb_utils.py`)

> **Objetivo:** Centralizar inicialização, login check e padrão de naming de runs no Weights & Biases. ([[docs.wandb.ai](https://docs.wandb.ai/guides/integrations/huggingface/?utm_source=chatgpt.com)][1])

1. **Implementar `init_wandb()`**

   * Parâmetros: `project_name`, `run_name`, `config` (dict de hiperparâmetros).
   * Chama `wandb.login()`; se falhar, imprime sugestão “wandb login” e aborta.
   * Chama `wandb.init(project=…, name=…, config=…)`.
2. **Testar fluxo de login**

   * Simular ambiente sem API key:

     * `init_wandb()` deve alertar `ERROR: please run 'wandb login'` e sair.
   * Com `WANDB_API_KEY` setado:

     * `init_wandb("proj", "run1", {})` retorna um objeto W\&B Run ativo.
3. **Acceptance Criteria:**

   * Wrapper detecta corretamente ausência de credenciais e avisa.
   * Após login, `wandb.run.id` está definido e visível em logs. ([[discuss.huggingface.co](https://discuss.huggingface.co/t/logging-experiment-tracking-with-w-b/498?utm_source=chatgpt.com)][2])

---

### 2.3 Pré-processamento de Dados (`utils/data_utils.py`)

> **Objetivo:** Oferecer funções genéricas para tokenizar lotes e empacotar batches dinamicamente com padding.

1. **Função `get_tokenizer()`**

   * Retorna o tokenizer do Qwen3 a partir dos arquivos `qwen3/config.json` e `qwen3/vocab.txt`.
2. **Função `tokenize_batch(texts: List[str], tokenizer)`**

   * Recebe lista de strings, retorna dict com `input_ids` e `attention_mask`.
3. **Função `get_data_collator(tokenizer)`**

   * Retorna `transformers.DataCollatorWithPadding(tokenizer=tokenizer)` para padding dinâmico. ([[huggingface.co](https://huggingface.co/docs/transformers/main_classes/data_collator?utm_source=chatgpt.com)][3])
4. **Acceptance Criteria:**

   * Num REPL:

     ```python
     tok = get_tokenizer()
     batch = tokenize_batch(["olá mundo", "teste"], tok)
     coll = get_data_collator(tok)
     padded = coll(batch)  # Deve conter tensors 'input_ids' e 'attention_mask' do mesmo comprimento
     ```
   * Não existem dependências diretas nos scripts de treino: usamos apenas `utils/data_utils.py`.

---

### 2.4 Smoke-Test e Registro Dummy em W\&B

> **Objetivo:** Validar **fim-a-fim** a integração de Hydra + W\&B + DataLoader antes de qualquer treino real.

1. **Criar `scripts/smoke_run.py`**

   * Usa `@hydra.main` para ler `default.yaml`.
   * Chama `init_wandb()` de `utils/wandb_utils.py`.
   * Carrega 2 strings de teste via `tokenize_batch()` e empacota com `get_data_collator()`.
   * Registra uma métrica dummy: `wandb.log({"smoke_loss": 0.123})`.
   * Finaliza com `wandb.finish()`.
2. **Acceptance Criteria:**

   * Executar `python scripts/smoke_run.py`:

     * Lê configurações sem erros.
     * Cria uma nova Run no painel W\&B com nome de projeto igual a `WANDB_PROJECT` ou configurado em Hydra.
     * Aparece no W\&B uma métrica `smoke_loss` igual a `0.123`.
     * Gera artefato de config (se `log_config=True`). ([[docs.wandb.ai](https://docs.wandb.ai/guides/integrations/huggingface/?utm_source=chatgpt.com)][1])

---

> **Ao final da Semana 2**, teremos um **sistema unificado de configuração** (Hydra), **logging padronizado** (W\&B wrapper) e **pipeline de dados reutilizável** (tokenização + DataCollator), com uma **execução de smoke-test** que comprova a integração de todas as peças.

[1]: https://docs.wandb.ai/guides/integrations/huggingface/?utm_source=chatgpt.com "Hugging Face Transformers | Weights & Biases Documentation"
[2]: https://discuss.huggingface.co/t/logging-experiment-tracking-with-w-b/498?utm_source=chatgpt.com "Logging & Experiment tracking with W&B - Transformers - Hugging Face ..."
[3]: https://huggingface.co/docs/transformers/main_classes/data_collator?utm_source=chatgpt.com "Data Collator - Hugging Face"
